Add quiver trace type for vector field visualization

degzhaus opened the pull request:
## Overview
This PR adds a new `quiver` trace type to Plotly.js for visualizing 2D vector fields using arrows.

## Features
- **Vector positioning**: Arrows placed at `x`, `y` coordinates with direction/magnitude from `u`, `v` components
- **Sizing modes**: `scaled` (normalized), `absolute`, or `raw` vector lengths via `sizemode`/`sizeref`
- **Arrow anchoring**: Position arrows by `tail`, `tip`, or `center`/`cm`/`middle`
- **Colorscale support**: Color arrows by magnitude or custom scalar field (`c` array)
- **Styling**: Configurable line color, width, dash patterns via `line.*` attributes
- **Arrowhead sizing**: Adjustable arrowhead size via `arrowsize`
- **Full interactivity**: Hover support with customizable `hovertemplate`, point selection
- **Animation support**: Animatable `x`, `y`, `u`, `v`, and `c` arrays

## API

```
Plotly.newPlot('div', [{
  type: 'quiver',
  x: [0, 1, 2],
  y: [0, 1, 2],
  u: [1, 0, -1],  // x-component of vectors
  v: [0, 1, 0],   // y-component of vectors
  sizemode: 'scaled',  // 'scaled' | 'absolute' | 'raw'
  sizeref: 0.5,
  anchor: 'tail',  // 'tail' | 'tip' | 'center'
  colorscale: 'Viridis',
  line: { width: 2 }
}]);
```

## Screenshots

Examples taken from [plotly.com/python/quiver-plots](https://plotly.com/python/quiver-plots/)

[Gist with example code](https://gist.github.com/degzhaus/2fb3c08cbc228682125f94bce52ba7d1)

### Basic Quiver Plot
<img width="724" alt="Basic quiver plot" src="https://github.com/user-attachments/assets/ff6ed199-a2a6-43ac-a789-37d4926f1bec" />

### With Colorscale
<img width="823" alt="Quiver with Viridis colorscale" src="https://github.com/user-attachments/assets/2f6697f4-6b07-4080-afdf-bb2b6771502e" />

## Testing
- ✅ Jasmine unit tests
- ✅ 9 visual regression test mocks

## Files Changed

### New files (`src/traces/quiver/`)
- `index.js` - Trace module definition
- `attributes.js` - Attribute schema
- `defaults.js` - Default value handling
- `calc.js` - Data calculation
- `plot.js` - SVG rendering
- `style.js` - Styling
- `hover.js` - Hover behavior
- `select_points.js` - Selection support
- `event_data.js` - Event data formatting
- `format_labels.js` - Label formatting

### Modified
- `lib/index.js` & `lib/index-strict.js` - Build integration

### Tests
- `test/jasmine/tests/quiver_test.js` - Unit tests
- `test/image/mocks/quiver_*.json` - 9 mock files
- `test/image/baselines/quiver_*.png` - Baseline images

degzhaus commented:
Hello, I am new to this project and attempting to contribute my first PR.

Unfortunately, I am struggling to get the build to pass. Specifically, the `publish-dist-node-v22` step:
https://app.circleci.com/pipelines/github/plotly/plotly.js/12658/workflows/ae5271e7-9971-4db1-a713-b42f0625901f/jobs/282198

I tried running `npm run schema` and pushing the change here:
https://github.com/plotly/plotly.js/pull/7584/commits/9463b3fa31c97d02896622569e7cc2dc3702e6b5

But still encountering the issue.

Any insight would be greatly appreciated. Thank you!

emilykl commented:
Hi @degzhaus, thank you for working on this contribution!

As it stands currently, the devtools dashboard needs to be running in order for `npm run schema` to work properly. So you should run `npm start` and leave it running, and then in another terminal run `npm run schema`. (@camdecoster is working on a PR to change that, see #7589 !)

You can also download `plot-schema.json` from the "Artifacts" tab in CircleCI and commit it.

emilykl commented:
@degzhaus One more comment — it would be great to include colorscale attributes in `quiver` so that arrows can be colored according to the magnitude of the vector. These attributes are added to the cone trace [here](https://github.com/plotly/plotly.js/blob/9e9572be2f1f0eaee24467037df5a5a731c5cffb/src/traces/cone/attributes.js#L168); something similar can be done for the quiver trace. And then obviously those attributes need to be referenced in the plotting code to assign the appropriate color.

gpdf commented:
@emilykl Would it be "Plotlyesque" for there to be an option to show arrows as "`o------`" as a selectable alternative to "`------>`"?  That is, instead of a line at the start and an arrowhead at the end, a dot at the start and a bare line at the end?  I've tried reading the existing arrow documentation and I'm not entirely clear whether this is currently part of the Plotly.js API.

gpdf commented:
> @gpdf The arrowhead attributes are poorly documented, but this is possible using the existing API — for an annotation, you can set `arrowside: 'start'` to show an arrowhead only at the start of the line, and `startarrowhead: 6` to make it shaped like a dot.

That's great.  That's another good argument to use the native arrow API rather than the "manual line drawing" approach from the Plotly-Python figure-factory function.

degzhaus commented:
> > @degzhaus One more comment — it would be great to include colorscale attributes in `quiver` so that arrows can be colored according to the magnitude of the vector. These attributes are added to the cone trace [here](https://github.com/plotly/plotly.js/blob/9e9572be2f1f0eaee24467037df5a5a731c5cffb/src/traces/cone/attributes.js#L168); something similar can be done for the quiver trace. And then obviously those attributes need to be referenced in the plotting code to assign the appropriate color.
> 
> I would suggest that this be capable of using any additional attribute of the data (e.g., another parallel data series), not just the magnitude. It can be very useful to color the vectors according to some independent state variable.

Thank you for this idea, @gpdf ! Added here:
 * https://github.com/plotly/plotly.js/pull/7584/commits/9ff8120b65646c26a2dc0cd5e1316ff47e6484aa

emilykl commented:
Great news @degzhaus, I've started looking at the code and I'll get back to you soon with another round of review.

One quick question -- your PR description refers to some Jasmine test files and image mocks which I don't see in the diff, did you maybe forget to push some files?

degzhaus commented:
> > > @degzhaus One more comment — it would be great to include colorscale attributes in `quiver` so that arrows can be colored according to the magnitude of the vector. These attributes are added to the cone trace [here](https://github.com/plotly/plotly.js/blob/9e9572be2f1f0eaee24467037df5a5a731c5cffb/src/traces/cone/attributes.js#L168); something similar can be done for the quiver trace. And then obviously those attributes need to be referenced in the plotting code to assign the appropriate color.
> > 
> > 
> > Amazing, thank you so much @emilykl !
> > I did indeed forget to push the tests, which i have done now here: [12a46e6](https://github.com/plotly/plotly.js/commit/12a46e631cb11aa025a53b9a914660b63c9d3185)
> > One thing I'm spinning my wheels on now is the image generation. I generated them locally here: [9287e5c](https://github.com/plotly/plotly.js/commit/9287e5c7b643f1c3cde3996874666253fc59032a)
> > But as you can see here, they are failing in ci: https://app.circleci.com/pipelines/github/plotly/plotly.js/12822/workflows/a9ae7478-acf1-4d50-817c-3e7fbb34e8c3/jobs/286792/artifacts
> > One notable strangeness is that they look good in my commit, but look strange in the ci artifacts. One example: <img alt="image" width="900" height="400" src="https://private-user-images.githubusercontent.com/111590/515361129-bc616324-25c5-4ea5-a587-a37f0467dea2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM0ODE2MDIsIm5iZiI6MTc2MzQ4MTMwMiwicGF0aCI6Ii8xMTE1OTAvNTE1MzYxMTI5LWJjNjE2MzI0LTI1YzUtNGVhNS1hNTg3LWEzN2YwNDY3ZGVhMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTExOFQxNTU1MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMDc1NjQ0MzNkMjk3MTliYTY4YTUwMDQzOGE5MDhkYzUyMThmYzYwNzE4ZWU5NDJmMDRlYmY0MTIyM2I3MjI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.r830peRklB3ozzVqPfM5zj7iIsE8smAI3TuoV_dQzlE">
> > Thanks again for all your help with this PR!
> 
> Okay, I see what is happening now, @emilykl !
> 
> I was looking at the diff, not the raw image that CI generates. Here is an example:
> 
> CI generates the following image for `diff-quiver_simple.png`: <img alt="image" width="500" height="500" src="https://private-user-images.githubusercontent.com/111590/515815884-248bed98-d000-45f9-a94c-691a46c18cde.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM1MzQ5NDUsIm5iZiI6MTc2MzUzNDY0NSwicGF0aCI6Ii8xMTE1OTAvNTE1ODE1ODg0LTI0OGJlZDk4LWQwMDAtNDVmOS1hOTRjLTY5MWE0NmMxOGNkZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTExOVQwNjQ0MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MWI0MGU2Y2MwOTFiYmEyNzlmNDc1NjdmOTVmNzE1NjY4NDAwNGU4OWNlZTkzYTY2NGFiODhiMTJhZTJmZGQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vT_GTC99n1WkBPGhY_9K4QBWZ-pXwbJZCZa3npWVJRQ">
> 
> And corresponding diff for the same image: <img alt="image" width="500" height="500" src="https://private-user-images.githubusercontent.com/111590/515816024-3ec00bd2-19b6-4a2a-aa1b-7b633ab714d4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM1MzQ5NDUsIm5iZiI6MTc2MzUzNDY0NSwicGF0aCI6Ii8xMTE1OTAvNTE1ODE2MDI0LTNlYzAwYmQyLTE5YjYtNGEyYS1hYTFiLTdiNjMzYWI3MTRkNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTExOVQwNjQ0MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNWFmY2JkNGJkZWYyNDNiYzZkMzE5MjdkZDkyZWEzNDNjZjQ2ZGMyYWUyY2ZlMDgzZGM3YWM5NGI1MzI4YmVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Fo9o85sEDEtCpcE5IppUB40durEKRyKatcSZHwp1VbY">
> 
> So, maybe something is different with my local versioning for dependencies?
> 
> My only fear is that if I swap in the raw images that CI generates, that it is actually wrong and/or could break or flake in the future. Hmm 🤔
> 
> Thanks again for everything, and I hope you have a splendid day!

@emilykl I got it 🥳 I hadn't worked with the `baselines` before, so took a little getting used to.

So excited! I will clean up some commits in the morning.

Thanks again!

degzhaus commented:
Pardon the delay, @emilykl, i was finally able to make time to clean this up. I removed some of my recent comments/questions because they were noise around me understanding how to work with the `tests/images`, but i was able to figure it out.

Thanks again for all your help with this PR! Excited for next steps.

degzhaus commented:
> Perhaps for future readers' clarity of understanding, @degzhaus , you could edit the description at the top of the PR to say:
> 
> ~scatterquiver~ `quiver`
> 
> in the two places it appears, so that the as-merged plot type appears there.

Great catch, thank you so much, @gpdf ! 

Updated per your recommendation, and did a little review/revision based on changes made during the review process as well.

degzhaus commented:
> Hi @degzhaus ! I'm aiming to give this another review by the end of the week. Sorry for the delay in getting back to you — looking forward to it! Thanks for your work so far.

Hey there, @emilykl, hope you're having a wonderful week!

I rebased, cleaned up some code, and added some tests here:
https://github.com/plotly/plotly.js/pull/7584/changes/0252e03b7827cf5c386056eb47f3c0bf4168c1de

Thanks again for everything!

emilykl commented:
Thanks @degzhaus ! Thanks for the heads up on the closed PR, I'm not sure exactly why that happened but I've linked this PR to the new one you've opened.
